README.rst
pan-iot-security-python
=======================

``pan-iot-security-python`` is a Python package for the Palo Alto
Networks
`IoT Security API
<https://docs.paloaltonetworks.com/iot/iot-security-api-reference.html>`_.

It provides a Python3
`asyncio <https://docs.python.org/3/library/asyncio.html>`_
and non-asyncio
class and command line interface to the IoT Security RESTful API.

Documentation
-------------

- Rendered reStructuredText on GitHub:

  https://github.com/PaloAltoNetworks/pan-iot-security-python/tree/main/doc

Install ``pan-iot-security-python``
-----------------------------------

``pan-iot-security-python`` is available as a
`release
<https://github.com/PaloAltoNetworks/pan-iot-security-python/releases/>`_
on GitHub and as a
`package
<https://pypi.org/project/pan-iot-security-python/>`_
on PyPi.


HISTORY.rst
Release History
===============

0.3.0 (2022-05-16)
------------------

- Add -O option, an optimised version of -j for the get all methods,
  which does not place all the results in memory.

- Add policies_all() method.

- Fix missing profile(), policy() in method attributes documentation.

- Unit test enhancements.

0.2.0 (2022-04-23)
------------------

- Add support for new API requests:

  - Get device profiles
  - Get policy rule recommendations

- Print stack trace entries in addition to exception information
  depending on debug:

  - debug 0: exception information
  - debug 1: stack trace limit=-1
  - debug > 1: stack trace limit=None

- /tag/list API request uses offset and pagelength.

- iotapy.py: Fix for missing await.

0.1.0 (2022-04-09)
------------------

- Change generator methods to return a tuple.  NOTE: not backward
  compatible.  See documentation for the 3 ``*_all()`` methods.

- Documentation fixes and improvements.

- Unit test fixes and improvements.

0.0.0 (2022-03-23)
------------------

- Initial release.


tests/mixin.py
import asyncio
import json
import logging
import os

import paniot


class _MixinShared:
    def iotapi(self):
        path = os.getenv('PANIOT_KEYS')
        if path is None:
            raise RuntimeError('no PANIOT_KEYS in environment')
        with open(path, 'r') as f:
            x = json.load(f)
        kwargs = {
            'customerid': x['customerid'],
            'access_key_id': x['access-key-id'],
            'access_key': x['access-key'],
        }

        x = os.getenv('PANIOT_DEBUG')
        if x is not None:
            debug = int(x)
            logger = logging.getLogger()
            if debug == 3:
                logger.setLevel(paniot.DEBUG3)
            elif debug == 2:
                logger.setLevel(paniot.DEBUG2)
            elif debug == 1:
                logger.setLevel(paniot.DEBUG1)
            elif debug == 0:
                pass
            else:
                raise RuntimeError('PANIOT_DEBUG level must be 0-3')

            log_format = '%(message)s'
            handler = logging.StreamHandler()
            formatter = logging.Formatter(log_format)
            handler.setFormatter(formatter)
            logger.addHandler(handler)

        return paniot.IotApi(**kwargs)


class Mixin(_MixinShared):
    def setUp(self):
        self.api = self.iotapi()

    def tearDown(self):
        self.api.session.close()


class AioMixin(_MixinShared):
    async def asyncSetUp(self):
        self.api = self.iotapi()

    async def asyncTearDown(self):
        await self.api.session.close()
        # XXX try to avoid "ResourceWarning: unclosed ..."
        await asyncio.sleep(0.1)


tests/test_sort.py
import unittest

from . import mixin


class IotApiTest(mixin.AioMixin, unittest.IsolatedAsyncioTestCase):
    async def test_01(self):
        x = {
            'sortfield': 'x-invalid',
        }
        resp = await self.api.device(query_string=x)
        self.assertEqual(resp.status, 400)

    async def test_02(self):
        key = 'MAC'
        x = {
            'sortfield': key,
            # default: desc
        }
        resp = await self.api.device(detail=True, query_string=x)
        self.assertEqual(resp.status, 200)
        x = await resp.json()
        last = 'ff:ff:ff:ff:ff:ff'
        for device in x['devices']:
            current = device[key]
            if last < current:
                self.fail('sort desc %s: %s < %s' % (key, last, current))
            last = current

    async def test_03(self):
        key = 'MAC'
        x = {
            'sortfield': key,
            'sortdirection': 'desc',
        }
        resp = await self.api.device(detail=True, query_string=x)
        self.assertEqual(resp.status, 200)
        x = await resp.json()
        last = 'ff:ff:ff:ff:ff:ff'
        for device in x['devices']:
            current = device[key]
            if last < current:
                self.fail('sort desc %s: %s < %s' % (key, last, current))
            last = current

    async def test_04(self):
        key = 'MAC'
        x = {
            'sortfield': key,
            'sortdirection': 'asc',
        }
        resp = await self.api.device(detail=True, query_string=x)
        self.assertEqual(resp.status, 200)
        x = await resp.json()
        last = '00:00:00:00:00:00'
        for device in x['devices']:
            # XXX when MAC is nil, key not in response
            # all nil MAC fields will be first for ascending
            if key in device:
                current = device[key]
                if last > current:
                    self.fail('sort asc %s: %s > %s' % (key, last, current))
                last = current

    async def test_05(self):
        x = {
            'sortfield': 'x-invalid',
        }
        resp = await self.api.alert(query_string=x)
        self.assertEqual(resp.status, 400)

    async def test_06(self):
        key = 'date'
        x = {
            'sortfield': key,
            # default: desc
        }
        resp = await self.api.alert(query_string=x)
        self.assertEqual(resp.status, 200)
        x = await resp.json()
        last = '9999'
        for item in x['items']:
            current = item[key]
            if last < current:
                self.fail('sort desc %s: %s < %s' % (key, last, current))
            last = current

    async def test_07(self):
        key = 'date'
        x = {
            'sortfield': key,
            'sortdirection': 'desc',
        }
        resp = await self.api.alert(query_string=x)
        self.assertEqual(resp.status, 200)
        x = await resp.json()
        last = '9999'
        for item in x['items']:
            current = item[key]
            if last < current:
                self.fail('sort desc %s: %s < %s' % (key, last, current))
            last = current

    async def test_08(self):
        key = 'date'
        x = {
            'sortfield': key,
            'sortdirection': 'asc',
        }
        resp = await self.api.alert(query_string=x)
        self.assertEqual(resp.status, 200)
        x = await resp.json()
        last = '0000'
        for item in x['items']:
            current = item[key]
            if last > current:
                self.fail('sort asc %s: %s > %s' % (key, last, current))
            last = current


tests/test_aio_tag.py
import unittest

from . import mixin


class IotApiTest(mixin.AioMixin, unittest.IsolatedAsyncioTestCase):
    async def test_01(self):
        x = {'source': 'x-invalid'}
        resp = await self.api.tag(query_string=x)
        self.assertEqual(resp.status, 400)

    async def test_02(self):
        resp = await self.api.tag()
        self.assertEqual(resp.status, 200)
        x = await resp.json()
        if x['totalTags'] <= 1000:
            self.assertEqual(x['totalTags'], len(x['tags']))
        if x['totalTags'] > 0:
            resp = await self.api.tag(pagelength=1)
            self.assertEqual(resp.status, 200)
            x = await resp.json()
            self.assertEqual(len(x['tags']), 1)
            if x['totalTags'] > 1:
                tag0 = x['tags'][0]
                resp = await self.api.tag(offset=0,
                                          pagelength=2)
                self.assertEqual(resp.status, 200)
                x = await resp.json()
                self.assertEqual(len(x['tags']), 2)
                self.assertEqual(x['tags'][0], tag0)
                tag1 = x['tags'][1]
                resp = await self.api.tag(offset=1,
                                          pagelength=1)
                self.assertEqual(resp.status, 200)
                x = await resp.json()
                self.assertEqual(len(x['tags']), 1)
                self.assertEqual(x['tags'][0], tag1)


paniot/mixin.py
#
# Copyright (c) 2022 Palo Alto Networks, Inc.
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

import aiohttp
import base64
import json
import logging
import requests
import requests.adapters
import ssl

from . import ArgsError, DEBUG1, DEBUG2, DEBUG3


class _MixinShared:
    def _auth(self, access_key_id, access_key):
        if access_key_id is None:
            raise ArgsError('access_key_id required')
        if access_key is None:
            raise ArgsError('access_key required')

        return {
            'X-Key-Id': access_key_id,
            'X-Access-Key': access_key,
        }

    def _method_attributes(self):
        METHOD_ATTRIBUTES = [
            (self.device, {'window': 60, 'rate_limit': 60}),
            (self.device_details, {'window': 60, 'rate_limit': 180}),
            (self.vulnerability, {'window': 60, 'rate_limit': 180}),
            (self.alert, {'window': 60, 'rate_limit': 180}),
            (self.tag, {'window': 60, 'rate_limit': 180}),
            (self.profile, {'window': 60, 'rate_limit': 180}),
            (self.policy, {'window': 60, 'rate_limit': 180}),
            (self.device_update, {'window': 60, 'rate_limit': 180}),
            (self.vuln_update, {'window': 60, 'rate_limit': 180}),
            (self.alert_update, {'window': 60, 'rate_limit': 180}),
        ]

        for x in METHOD_ATTRIBUTES:
            func, attrs = x
            func.__func__.window = attrs['window']
            func.__func__.rate_limit = attrs['rate_limit']

    def decode_jwt(self):
        def _pad(x):
            remainder = len(x) % 4
            if remainder > 0:
                x += b'=' * (4 - remainder)
            return x

        try:
            jwt = self.jwt.encode().split(b'.')
            if len(jwt) != 3:
                raise ArgsError('token is not a JSON Web Signature')
            x = base64.urlsafe_b64decode(_pad(jwt[0]))
            header = json.loads(x)
            if 'typ' not in header or header['typ'] != 'JWT':
                raise ArgsError('token is not a JSON Web Token')
            x = base64.urlsafe_b64decode(_pad(jwt[1]))
            payload = json.loads(x)
        except (base64.binascii.Error, ValueError) as e:
            raise ArgsError(e)

        return header, payload


class AioMixin(_MixinShared):
    async def __aenter__(self):
        self._log(DEBUG2, '%s', '__aenter__')
        return self

    async def __aexit__(self, *args):
        self._log(DEBUG2, '%s', '__aexit__')
        if not self.session.closed:
            self._log(DEBUG1, 'closing aiohttp session')
            await self.session.close()

    def _timeout(self, timeout):
        if timeout is None:
            return

        if isinstance(timeout, tuple):
            if len(timeout) != 2:
                raise ArgsError('timeout tuple length must be 2')
            x = aiohttp.ClientTimeout(sock_connect=timeout[0],
                                      sock_read=timeout[1])
        else:
            x = aiohttp.ClientTimeout(total=timeout)

        return x

    def _ssl_context(self, verify):
        context = ssl.create_default_context(purpose=ssl.Purpose.SERVER_AUTH)

        if isinstance(verify, bool):
            if not verify:
                context.check_hostname = False
                context.verify_mode = ssl.CERT_NONE
        elif verify is not None:
            try:
                context.load_verify_locations(cafile=verify)
            except (FileNotFoundError, ssl.SSLError) as e:
                raise ValueError('%s: %s' % (verify, e))

        return context

    def _session(self, auth=None, timeout=None):
        async def on_request_start(session, trace_config_ctx, params):
            log = logging.getLogger(__name__).log
            log(DEBUG2, '%s %s', params.method, params.url)
            for k, v in params.headers.items():
                x = '*' * 6 if k == 'X-Access-Key' else v
                log(DEBUG3, '%s: %s', k, x)

        async def on_request_chunk_sent(session, trace_config_ctx, params):
            log = logging.getLogger(__name__).log
            if params.chunk:
                log(DEBUG3, '%s', params.chunk)

        async def on_request_end(session, trace_config_ctx, params):
            log = logging.getLogger(__name__).log
            log(DEBUG1, '%s %s %s %s %s',
                params.method,
                params.url,
                params.response.status,
                params.response.reason,
                params.response.headers.get('content-length'))
            for k, v in params.response.headers.items():
                log(DEBUG3, '%s: %s', k, v)

        kwargs = {}
        if auth is not None:
            kwargs['headers'] = auth
        if timeout is not None:
            kwargs['timeout'] = timeout

        if (logging.getLogger(__name__).getEffectiveLevel() in
           [DEBUG1, DEBUG2, DEBUG3]):
            trace_config = aiohttp.TraceConfig()
            trace_config.on_request_start.append(on_request_start)
            trace_config.on_request_chunk_sent.append(on_request_chunk_sent)
            trace_config.on_request_end.append(on_request_end)
            kwargs['trace_configs'] = [trace_config]

        return aiohttp.ClientSession(**kwargs)


class _TimeoutHTTPAdapter(requests.adapters.HTTPAdapter):
    def __init__(self, *args, **kwargs):
        self.timeout = None
        if 'timeout' in kwargs:
            self.timeout = kwargs['timeout']
            del kwargs['timeout']
        super().__init__(*args, **kwargs)

    def send(self, request, **kwargs):
        timeout = kwargs.get('timeout')
        if timeout is None:
            kwargs['timeout'] = self.timeout
        return super().send(request, **kwargs)


class Mixin(_MixinShared):
    def __enter__(self):
        self._log(DEBUG2, '%s', '__enter__')
        return self

    def __exit__(self, *args):
        self._log(DEBUG2, '%s', '__exit__')
        self._log(DEBUG1, 'closing requests session')
        self.session.close()

    def _session(self,
                 auth=None,
                 verify=None,
                 timeout=None):
        session = requests.Session()

        if auth is not None:
            session.headers.update(auth)
        if verify is not None:
            session.verify = verify
        if timeout is not None:
            adapter = _TimeoutHTTPAdapter(timeout=timeout)
            session.mount("https://", adapter)

        return session


paniot/__init__.py
#
# Copyright (c) 2022 Palo Alto Networks, Inc.
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

import asyncio
from collections import namedtuple
import logging
import re
import sys

__version__ = '0.3.0'

_default_api_version = (4, 0)
DEFAULT_API_VERSION = 'v%d.%d' % _default_api_version

DEBUG1 = logging.DEBUG
DEBUG2 = DEBUG1 - 1
DEBUG3 = DEBUG2 - 1

logging.addLevelName(DEBUG2, 'DEBUG2')
logging.addLevelName(DEBUG3, 'DEBUG3')


class ApiError(Exception):
    pass


class ArgsError(ApiError):
    pass


class ApiVersion(namedtuple('api_version',
                            ['major', 'minor'])):
    def __str__(self):
        return 'v%d.%d' % (self.major, self.minor)

    def __int__(self):
        # reserve lower 8 bits for 'future' use
        return self.major << 16 | self.minor << 8


def _isaio():
    try:
        asyncio.get_running_loop()
        return True
    except RuntimeError:
        return False


def IotApi(api_version=None, *args, **kwargs):
    _log = logging.getLogger(__name__).log

    if api_version is None:
        x = _default_api_version
    else:
        r = re.search(r'^v?(\d+)\.(\d+)$', api_version)
        if r is None:
            raise ArgsError('Invalid api_version: %s' % api_version)
        x = int(r.group(1)), int(r.group(2))
    _api_version = ApiVersion(*x)
    _log(DEBUG1, 'api_version: %s, 0x%06x',
         _api_version, int(_api_version))

    package = 'paniot'
    name = 'aioapi' if _isaio() else 'api'
    module = 'v%d_%d%s' % (_api_version.major,
                           _api_version.minor,
                           name)
    module_name = '%s.%s' % (package, module)
    class_ = 'IotApi'

    try:
        __import__(module_name)
    except ImportError as e:
        raise ArgsError('Module import error: %s: %s' %
                        (module_name, e))

    try:
        klass = getattr(sys.modules[module_name], class_)
    except AttributeError:
        raise ArgsError('Class not found: %s' % class_)

    return klass(api_version=_api_version, *args, **kwargs)


paniot/v4_0aioapi.py
#
# Copyright (c) 2022 Palo Alto Networks, Inc.
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

import aiohttp
import asyncio
import logging
import ssl
import sys

from . import (mixin, ApiError, ArgsError,
               DEBUG1, DEBUG2, DEBUG3, __version__)


BASE_PATH = '/pub/v4.0'


class IotApi(mixin.AioMixin):
    def __init__(self, *,
                 api_version=None,
                 url=None,
                 access_key_id=None,
                 access_key=None,
                 customerid=None,
                 verify=None,
                 timeout=None):
        self._log = logging.getLogger(__name__).log
        self._log(DEBUG2, 'pan-iot-security-python: %s, IotApi: %s',
                  __version__, api_version)
        self._log(DEBUG2, 'Python version: %s', sys.version)
        self._log(DEBUG2, 'ssl: %s', ssl.OPENSSL_VERSION)
        self._log(DEBUG2, 'aiohttp: %s', aiohttp.__version__)

        self.api_version = api_version
        if customerid is None:
            raise ArgsError('customerid required')
        self.customerid = customerid
        if url is None:
            self.url = 'https://%s.iot.paloaltonetworks.com' % customerid
        else:
            self.url = url
        try:
            self.ssl = self._ssl_context(verify)
        except ValueError as e:
            raise ArgsError(e)
        self._log(DEBUG2, 'ssl: %s %s', self.ssl.verify_mode,
                  self.ssl.check_hostname)
        auth = self._auth(access_key_id, access_key)
        self.jwt = access_key
        timeout_ = self._timeout(timeout)
        self._log(DEBUG2, 'timeout: %s', timeout_)
        self.session = self._session(auth=auth, timeout=timeout_)
        self._method_attributes()

    async def _request_retry(self, *,
                             retry=False,
                             method=None,
                             retry_timeout=False,
                             func=None,
                             **kwargs):
        if retry:
            assert method is not None, 'method required when retry'
            rate_limit_delay = method.window / method.rate_limit
        if retry_timeout:
            timeout_delay = 5
            timeout_retries = 3

        while True:
            try:
                resp = await func(**kwargs)
            except asyncio.TimeoutError:
                if not (retry_timeout and timeout_retries):
                    raise
                self._log(DEBUG2, 'timeout, sleep %.2fs', timeout_delay)
                await asyncio.sleep(timeout_delay)
                timeout_delay *= 2
                timeout_retries -= 1
            else:
                if retry and resp.status == 429:
                    self._log(DEBUG2, 'status code 429, sleep %.2fs',
                              rate_limit_delay)
                    await asyncio.sleep(rate_limit_delay)
                    if rate_limit_delay < method.window / 4:
                        rate_limit_delay *= 2
                else:
                    break

        return resp

    async def device(self, *,
                     stime=None,
                     detail=False,
                     offset=None,
                     pagelength=None,
                     query_string=None,
                     retry=False):
        path = BASE_PATH + '/device/list'
        url = self.url + path

        params = {'customerid': self.customerid}
        if stime is not None:
            params['stime'] = stime
        if detail is True:
            params['detail'] = 'true'
        if offset is not None:
            params['offset'] = offset
        if pagelength is not None:
            params['pagelength'] = pagelength
        if query_string is not None:
            params.update(query_string)

        kwargs = {
            'url': url,
            'ssl': self.ssl,
            'params': params,
        }

        resp = await self._request_retry(retry=retry,
                                         method=self.device,
                                         func=self.session.get,
                                         **kwargs)

        return resp

    async def _get_all(self, func, keys, **kwargs):
        def _get(x, keys):
            if not keys:
                return x
            return _get(x[keys[0]], keys[1:])

        offset = 0
        pagelength = 1000

        while True:
            resp = await func(offset=offset,
                              pagelength=pagelength,
                              retry=True,
                              **kwargs)
            if resp.status == 200:
                obj = await resp.json(content_type=None)
                try:
                    obj = _get(obj, keys)
                except KeyError as e:
                    raise ApiError('Malformed response, missing key %s' % e)
                length = len(obj)
                self._log(DEBUG2, 'length %d', length)
                for x in obj:
                    yield True, x

                if length < pagelength:
                    self._log(DEBUG1, 'total %d', offset+length)
                    break
                offset += length
            else:
                yield False, resp

    async def devices_all(self, *,
                          stime=None,
                          detail=False,
                          query_string=None):
        kwargs = {
            'stime': stime,
            'detail': detail,
            'query_string': query_string,
        }

        async for x in self._get_all(func=self.device,
                                     keys=['devices'],
                                     **kwargs):
            yield x

    async def device_details(self, *,
                             deviceid=None,
                             ip=None,
                             query_string=None,
                             retry=False):
        if deviceid is None and ip is None:
            raise ArgsError('deviceid or ip required')
        if deviceid is not None and ip is not None:
            raise ArgsError(
                'deviceid and ip cannot be used at the same time')

        kwargs = {
            'ssl': self.ssl,
            'params': {'customerid': self.customerid},
        }

        if deviceid is not None:
            path = BASE_PATH + '/device'
            url = self.url + path
            kwargs['params']['deviceid'] = deviceid
        if ip is not None:
            path = BASE_PATH + '/device/ip'
            url = self.url + path
            kwargs['params']['ip'] = ip
        if query_string is not None:
            kwargs['params'].update(query_string)

        kwargs['url'] = url

        resp = await self._request_retry(retry=retry,
                                         method=self.device_details,
                                         func=self.session.get,
                                         **kwargs)

        return resp

    async def vulnerability(self, *,
                            groupby=None,
                            stime=None,
                            deviceid=None,
                            offset=None,
                            pagelength=None,
                            query_string=None,
                            retry=False):
        path = BASE_PATH + '/vulnerability/list'
        url = self.url + path

        params = {'customerid': self.customerid}
        if groupby is None:
            params['groupby'] = 'vulnerability'
        else:
            params['groupby'] = groupby
        if stime is not None:
            params['stime'] = stime
        if deviceid is not None:
            params['deviceid'] = deviceid
        if offset is not None:
            params['offset'] = offset
        if pagelength is not None:
            params['pagelength'] = pagelength
        if query_string is not None:
            params.update(query_string)

        kwargs = {
            'url': url,
            'ssl': self.ssl,
            'params': params,
        }

        resp = await self._request_retry(retry=retry,
                                         method=self.vulnerability,
                                         func=self.session.get,
                                         **kwargs)

        return resp

    async def vulnerabilities_all(self, *,
                                  groupby=None,
                                  stime=None,
                                  query_string=None):
        kwargs = {
            'groupby': groupby,
            'stime': stime,
            'query_string': query_string,
        }
        keys = ['items']
        if groupby is None or groupby == 'vulnerability':
            keys.append('items')

        async for x in self._get_all(func=self.vulnerability,
                                     keys=keys,
                                     **kwargs):
            yield x

    async def alert(self, *,
                    stime=None,
                    offset=None,
                    pagelength=None,
                    query_string=None,
                    retry=False):
        path = BASE_PATH + '/alert/list'
        url = self.url + path

        params = {'customerid': self.customerid}
        if stime is not None:
            params['stime'] = stime
        if offset is not None:
            params['offset'] = offset
        if pagelength is not None:
            params['pagelength'] = pagelength
        if query_string is not None:
            params.update(query_string)

        kwargs = {
            'url': url,
            'ssl': self.ssl,
            'params': params,
        }

        resp = await self._request_retry(retry=retry,
                                         method=self.alert,
                                         func=self.session.get,
                                         **kwargs)

        return resp

    async def alerts_all(self, *,
                         stime=None,
                         query_string=None):
        kwargs = {
            'stime': stime,
            'query_string': query_string,
        }

        async for x in self._get_all(func=self.alert,
                                     keys=['items'],
                                     **kwargs):
            yield x

    async def tag(self, *,
                  offset=None,
                  pagelength=None,
                  query_string=None,
                  retry=False):
        path = BASE_PATH + '/tag/list'
        url = self.url + path

        params = {
            'customerid': self.customerid,
            'source': 'tenant',
        }
        if offset is not None:
            params['offset'] = offset
        if pagelength is not None:
            params['pagelength'] = pagelength
        if query_string is not None:
            params.update(query_string)

        kwargs = {
            'url': url,
            'ssl': self.ssl,
            'params': params,
        }

        resp = await self._request_retry(retry=retry,
                                         method=self.tag,
                                         func=self.session.get,
                                         **kwargs)

        return resp

    async def profile(self, *,
                      query_string=None,
                      retry=False):
        path = BASE_PATH + '/profile/mapping'
        url = self.url + path

        params = {
            'customerid': self.customerid,
        }
        if query_string is not None:
            params.update(query_string)

        kwargs = {
            'url': url,
            'ssl': self.ssl,
            'params': params,
        }

        resp = await self._request_retry(retry=retry,
                                         method=self.tag,
                                         func=self.session.get,
                                         **kwargs)

        return resp

    async def policy(self, *,
                     offset=None,
                     pagelength=None,
                     query_string=None,
                     retry=False):
        path = BASE_PATH + '/policy/recommendation'
        url = self.url + path

        params = {
            'customerid': self.customerid,
        }
        if offset is not None:
            params['offset'] = offset
        if pagelength is not None:
            params['pagelength'] = pagelength
        if query_string is not None:
            params.update(query_string)

        kwargs = {
            'url': url,
            'ssl': self.ssl,
            'params': params,
        }

        resp = await self._request_retry(retry=retry,
                                         method=self.tag,
                                         func=self.session.get,
                                         **kwargs)

        return resp

    async def policies_all(self, *,
                           query_string=None):
        kwargs = {
            'query_string': query_string,
        }

        async for x in self._get_all(func=self.policy,
                                     keys=['policies'],
                                     **kwargs):
            yield x

    async def device_update(self, *,
                            json=None,
                            query_string=None,
                            retry=False):
        path = BASE_PATH + '/device/update'
        url = self.url + path

        params = {
            'customerid': self.customerid,
        }
        if query_string is not None:
            params.update(query_string)

        kwargs = {
            'url': url,
            'ssl': self.ssl,
            'params': params,
        }
        if json is not None:
            if isinstance(json, (bytes, str, bytearray)):
                kwargs['data'] = json
                kwargs['headers'] = {'content-type': 'application/json'}
            else:
                kwargs['json'] = json

        resp = await self._request_retry(retry=retry,
                                         method=self.device_update,
                                         func=self.session.put,
                                         **kwargs)

        return resp

    async def vuln_update(self, *,
                          json=None,
                          query_string=None,
                          retry=False):
        path = BASE_PATH + '/vulnerability/update'
        url = self.url + path

        params = {
            'customerid': self.customerid,
        }
        if query_string is not None:
            params.update(query_string)

        kwargs = {
            'url': url,
            'ssl': self.ssl,
            'params': params,
        }
        if json is not None:
            if isinstance(json, (bytes, str, bytearray)):
                kwargs['data'] = json
                kwargs['headers'] = {'content-type': 'application/json'}
            else:
                kwargs['json'] = json

        resp = await self._request_retry(retry=retry,
                                         method=self.vuln_update,
                                         func=self.session.put,
                                         **kwargs)

        return resp

    async def alert_update(self, *,
                           id=None,
                           json=None,
                           query_string=None,
                           retry=False):
        path = BASE_PATH + '/alert/update'
        url = self.url + path

        params = {
            'customerid': self.customerid,
        }
        if id is not None:
            params['id'] = id
        if query_string is not None:
            params.update(query_string)

        kwargs = {
            'url': url,
            'ssl': self.ssl,
            'params': params,
        }
        if json is not None:
            if isinstance(json, (bytes, str, bytearray)):
                kwargs['data'] = json
                kwargs['headers'] = {'content-type': 'application/json'}
            else:
                kwargs['json'] = json

        resp = await self._request_retry(retry=retry,
                                         method=self.alert_update,
                                         func=self.session.put,
                                         **kwargs)

        return resp
